Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp

```diff
@@ -152,51 +170,149 @@ ActsExamples::ProcessCode ActsExamples::TrackFindingAlgorithm::execute(
+          Acts::smoothTrack(ctx.geoContext, firstTrack, logger());
+      if (!firstSmoothingResult.ok()) {
         m_nFailedSmoothing++;
-        ACTS_ERROR("Smoothing for seed "
-                   << iseed << " and track " << track.index()
-                   << " failed with error " << smoothingResult.error());
+        ACTS_ERROR("First smoothing for seed "
+                   << iSeed << " and track " << firstTrack.index()
+                   << " failed with error " << firstSmoothingResult.error());
         continue;
       }
 
-      auto extrapolationResult = Acts::extrapolateTrackToReferenceSurface(
-          track, *pSurface, extrapolator, extrapolationOptions,
-          m_cfg.extrapolationStrategy, logger());
-      if (!extrapolationResult.ok()) {
-        m_nFailedExtrapolation++;
-        ACTS_ERROR("Extrapolation for seed "
-                   << iseed << " and track " << track.index()
-                   << " failed with error " << extrapolationResult.error());
-        continue;
-      }
+      std::size_t nSecond = 0;
 
       // Set the seed number, this number decrease by 1 since the seed number
-      seedNumber(track) = nSeed - 1;
+      // has already been updated
+      seedNumber(firstTrack) = nSeed - 1;
+
+      if (m_cfg.twoWay) {
+        std::optional<Acts::VectorMultiTrajectory::TrackStateProxy> firstState;
+        for (auto st : firstTrack.trackStatesReversed()) {
+          bool isMeasurement =
+              st.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag);
+          bool isOutlier =
+              st.typeFlags().test(Acts::TrackStateFlag::OutlierFlag);
+          // We are excluding non measurement states and outlier here. Those can
+          // decrease resolution because only the smoothing corrected the very
+          // first prediction as filtering is not possible.
+          if (isMeasurement && !isOutlier) {
+            firstState = st;
+          }
+        }
+
+        if (firstState.has_value()) {
+          Acts::BoundTrackParameters secondInitialParameters(
+              firstState->referenceSurface().getSharedPtr(),
+              firstState->parameters(), firstState->covariance(),
+              firstInitialParameters.particleHypothesis());
+
+          auto secondResult = (*m_cfg.findTracks)(secondInitialParameters,
+                                                  secondOptions, tracksTemp);
+
+          if (!secondResult.ok()) {
+            ACTS_WARNING("Second track finding failed for seed "
+                         << iSeed << " with error" << secondResult.error());
+          } else {
+            auto firstFirstState =
+                std::next(firstTrack.trackStatesReversed().begin(),
+                          firstTrack.nTrackStates() - 1);
+
+            auto& secondTracksForSeed = secondResult.value();
+            for (auto& secondTrack : secondTracksForSeed) {
+              if (secondTrack.nTrackStates() < 2) {
+                continue;
```

Comment:
My thinking was that this will restore to the behavior without two way

Reply:
yes, that was my understanding. I was worried that `findTracks(secondInitialParameters,...)` could change the first track's parameters/covariance. That would be bad if it turns out we don't use the second track.

I guess you are saying that it isn't `findTracks` that can affect the whole track, but only the subsequent smoothing which is done once we are committed to the second track. Is that right?

Reply:
The second pass should not affect the `firstTrack`. So the track states should really stay the same. The final params are then determined using the extrapolation which will use one of the unchanged track states. So yes, the `findTracks` will not affect the whole track.

Reply:
Thanks for confirming. I guess this could still be an issue with the original all-in-one CKF (prior to #2722) that we still have to use in Athena. We don't have to worry about multiple branches (because we only use one branch at the moment), but maybe do for cases where the second track fails and we revert to the first - most common where we have < 2 track states. Is it worth trying to mitigate this? I thought to at least add the restoration of the original initial index as you do in this PR.

Reply:
We can take care of that when we bump the Acts version. I plan to release tomorrow and open a MR in Athena with the bumped version. Then we can check the two way implementation in Athena